File: 30. 包含min函数的栈.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.A, self.B = [], []


    def push(self, x: int) -> None:
        self.A.append(x)
        if not self.B or x<=self.B[-1]:
            self.B.append(x)

    def pop(self) -> None:
        if self.A.pop() == self.B[-1]:
            self.B.pop()

    def top(self) -> int:
        return self.A[-1]

    def min(self) -> int:
        return self.B[-1]

File: 剑指offer/C++_Version/30. 包含min函数的栈.cpp
//方法1
class MinStack {

    stack<int> A;
    stack<int> B;
    
public:
    /** initialize your data structure here. */
    MinStack()
    {
    }

    void push(int x) 
    {
        A.push(x);
        if(B.empty() || B.top()>=x) 
            B.push(x);
    }
    
    void pop() 
    {
        int ans = A.top();
        A.pop();
        if(B.top()==ans)
            B.pop();
    }
    
    int top() 
    {
        return A.top();
    }
    
    int min() 
    {
        return B.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->min();
 */
